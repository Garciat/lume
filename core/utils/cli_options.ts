import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  // Detect location and port
  let port: number;
  let location: URL;

  if (serveMode) {
    location = cli.location
      ? new URL(cli.location)
      : new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      location.port = port.toString();
    } else if (location.port) {
      port = parseInt(location.port);
    } else if (options.server?.port) {
      port = options.server.port;
      location.port = port.toString();
    } else {
      port = location.protocol === "https:" ? 443 : 3000;
      location.port = port.toString();
    }
  } else {
    location = cli.location
      ? new URL(cli.location)
      : (options.location as URL | undefined) || new URL("http://localhost");

    if (cli.port) {
      port = parseInt(cli.port);
      location.port = port.toString();
    } else if (location.port) {
      port = parseInt(location.port);
    } else {
      port = location.protocol === "https:" ? 443 : 80;
    }
  }

  options.location = location;
  options.server ||= {};
  options.server.port = port;

  if (cli.open) {
    options.server.open = cli.open;
  }

  return options;
}
