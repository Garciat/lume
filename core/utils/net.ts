export function localIp(): string | undefined {
  // Try/catch for https://github.com/denoland/deno/issues/25420
  try {
    for (const info of Deno.networkInterfaces()) {
      if (info.family !== "IPv4" || info.address.startsWith("127.")) {
        continue;
      }

      return info.address;
    }
  } catch {
    return undefined;
  }
}
