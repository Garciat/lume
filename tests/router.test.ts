import { assertResponseSnapshot, getServer } from "./utils.ts";
import Router from "../middlewares/router.ts";

Deno.test("simple routes", async (t) => {
  const server = getServer(() => new Response("Not Found", { status: 404 }));

  const router = new Router();

  router.get("/hello", () => new Response("Hello World from GET"));
  router.post("/hello", () => new Response("Hello World from POST"));
  router.put("/hello", () => new Response("Hello World from PUT"));
  router.delete("/hello", () => new Response("Hello World from DELETE"));

  server.use(router.middleware());

  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://example.com/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello", { method: "POST" }),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://example.com:1234/hello", { method: "POST" }),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello", { method: "PUT" }),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello", { method: "DELETE" }),
  );
});

Deno.test("simple routes with base path", async (t) => {
  const server = getServer(() => new Response("Not Found", { status: 404 }));

  const router = new Router({
    basePath: "/api",
  });

  router.get("/hello", () => new Response("Hello World from GET"));

  server.use(router.middleware());

  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://example.com/api/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://example.com:1234/api/hello"),
  );
});

Deno.test("routes with parameters", async (t) => {
  const server = getServer(() => new Response("Not Found", { status: 404 }));

  const router = new Router();

  router.get("/hello/:name", ({ name }) => new Response(`Hello ${name}`));

  server.use(router.middleware());

  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar/"),
  );
});

Deno.test("routes with parameters and strict = false", async (t) => {
  const server = getServer(() => new Response("Not Found", { status: 404 }));

  const router = new Router({
    strict: false,
  });

  router.get(
    "/hello/:name",
    ({ name, request }) =>
      new Response(
        `Hello ${name} / ${new URL(request.url).searchParams.get("query")}`,
      ),
  );

  server.use(router.middleware());

  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar/"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar/?query=foo"),
  );
  await assertResponseSnapshot(
    t,
    server,
    new Request("http://localhost/hello/oscar?query=foo"),
  );
});
