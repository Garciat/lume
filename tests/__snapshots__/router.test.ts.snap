export const snapshot = {};

snapshot[`simple routes 1`] = `
{
  body: "Hello World from GET",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes 2`] = `
{
  body: "Hello World from GET",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://example.com/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes 3`] = `
{
  body: "Hello World from POST",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "POST",
  request: "http://localhost/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes 4`] = `
{
  body: "Hello World from POST",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "POST",
  request: "http://example.com:1234/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes 5`] = `
{
  body: "Hello World from PUT",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "PUT",
  request: "http://localhost/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes 6`] = `
{
  body: "Hello World from DELETE",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "DELETE",
  request: "http://localhost/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes with base path 1`] = `
{
  body: "Not Found",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello",
  status: 404,
  statusText: "",
}
`;

snapshot[`simple routes with base path 2`] = `
{
  body: "Hello World from GET",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://example.com/api/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`simple routes with base path 3`] = `
{
  body: "Hello World from GET",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://example.com:1234/api/hello",
  status: 200,
  statusText: "",
}
`;

snapshot[`routes with parameters 1`] = `
{
  body: "Not Found",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello",
  status: 404,
  statusText: "",
}
`;

snapshot[`routes with parameters 2`] = `
{
  body: "Hello oscar",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar",
  status: 200,
  statusText: "",
}
`;

snapshot[`routes with parameters 3`] = `
{
  body: "Not Found",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar/",
  status: 404,
  statusText: "",
}
`;

snapshot[`routes with parameters and strict = false 1`] = `
{
  body: "Not Found",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello",
  status: 404,
  statusText: "",
}
`;

snapshot[`routes with parameters and strict = false 2`] = `
{
  body: "Hello oscar / null",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar",
  status: 200,
  statusText: "",
}
`;

snapshot[`routes with parameters and strict = false 3`] = `
{
  body: "Hello oscar / null",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar/",
  status: 200,
  statusText: "",
}
`;

snapshot[`routes with parameters and strict = false 4`] = `
{
  body: "Hello oscar / foo",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar/?query=foo",
  status: 200,
  statusText: "",
}
`;

snapshot[`routes with parameters and strict = false 5`] = `
{
  body: "Hello oscar / foo",
  headers: {
    "content-type": "text/plain;charset=UTF-8",
  },
  method: "GET",
  request: "http://localhost/hello/oscar?query=foo",
  status: 200,
  statusText: "",
}
`;
